### PR TITLE
feat: Send and process ACK-ECN

### PR DESCRIPTION
**Pull request by larseggert**
Most of the remaining bits from #1495

Depends on #1604, will retarget this at mozilla/neqo once that has landed

### CODE REVIEW

File: neqo-transport/src/connection/mod.rs [hunk 1]
```diff
@@ -2902,6 +2920,44 @@ impl Connection {
...
+            let newly_acked = acked_packets.len() as u64;
+            let ecn_diff = diff_ecn_count(&ecn_count, &path.borrow().ecn_count(space));
+            let sum_inc = ecn_diff[IpTosEcn::Ect0] + ecn_diff[IpTosEcn::Ce];
+            let cur_mark: IpTosEcn = path.borrow().tos().into();
+            if sum_inc < newly_acked {
+                qwarn!(
+                    "ACK had {} new marks, but sent {} packets, disabling ECN",
+                    ecn_diff[cur_mark],
+                    newly_acked
+                );
```
Comment (mxinden) [inline neqo-transport/src/connection/mod.rs]: Based on the enclosing `if` we know that ECN is enabled on the path. `ecn_diff[cur_mark]` ignores all CE marks. Wouldn't the following be correct?

```suggestion
                qwarn!(
                    "ACK had {} new marks, but acked {} packets, disabling ECN",
                    sum_inc,
                    newly_acked
                );
```

File: neqo-transport/src/tracking.rs [hunk 2]
```diff
@@ -397,9 +418,15 @@ impl RecvdPackets {
...
             unacknowledged_tolerance: DEFAULT_ACK_PACKET_TOLERANCE,
             ignore_order: false,
+            ecn_count: EcnCount::default(),
         }
     }
 
+    /// Increase the ECN count for the mark given by `ecn` by `n`.
+    pub fn inc_ecn_count(&mut self, ecn: IpTosEcn, n: u64) {
+        self.ecn_count[ecn] += n;
+    }
```
Comment (mxinden) [inline neqo-transport/src/tracking.rs]: Nit, given that `inc_ecn_count` is only ever called with `n = 1`:

```suggestion
    /// Increase the ECN count for the mark given by `ecn` by `1`.
    pub fn inc_ecn_count(&mut self, ecn: IpTosEcn) {
        self.ecn_count[ecn] += 1;
    }
```

File: neqo-transport/src/tracking.rs [hunk 3]
```diff
@@ -16,18 +16,20 @@ use std::{
...
 use smallvec::{smallvec, SmallVec};
 
 use crate::{
+    frame::{FRAME_TYPE_ACK, FRAME_TYPE_ACK_ECN},
     packet::{PacketBuilder, PacketNumber, PacketType},
     recovery::RecoveryToken,
     stats::FrameStats,
 };
 
 // TODO(mt) look at enabling EnumMap for this: https://stackoverflow.com/a/44905797/1375574
```
Comment (mxinden) [inline neqo-transport/src/tracking.rs:31]: With the addition below (`#[derive(..., Enum)]`) is this comment still necessary?

File: neqo-client/src/main.rs [hunk 4]
```diff
@@ -241,8 +241,7 @@ impl Args {
         // Only use v1 for most QNS tests.
         self.quic_parameters.quic_version = vec![Version::Version1];
         match testcase.as_str() {
-            // TODO: Add "ecn" when that is ready.
-            "http3" => {}
+            "http3" | "ecn" => {}
```
Comment (mxinden) [inline neqo-client/src/main.rs]: Missing corresponding change in `server/src/main.rs`:

``` diff
modified   neqo-server/src/main.rs
@@ -734,7 +734,7 @@ async fn main() -> Result<(), io::Error> {
         }
 
         match testcase.as_str() {
-            "http3" => (),
+            "http3" | "ecn" => (),
             "zerortt" => {
                 args.use_old_http = true;
                 args.alpn = String::from(HQ_INTEROP);
```

**Review by mxinden — COMMENTED** [review]
I have rebased the commits onto latest https://github.com/mozilla/neqo/pull/1604 and added a couple of smaller patches.

I have run this through the ECN QUIC Interop testcase. As of now, the server receives ECT0 from the client, but the server fails to properly send ECT0 to the client. This is due to wrong ECN handling of IPv4-Mapped IPv6 addresses. I will have a patch coming shortly. With the patch the ECN testcase between a neqo client and server succeeds.

//CC @larseggert to prevent duplicate work.

**Comment by mxinden** [comment]
> I have run this through the ECN QUIC Interop testcase. As of now, the server receives ECT0 from the client, but the server fails to properly send ECT0 to the client. This is due to wrong ECN handling of IPv4-Mapped IPv6 addresses. I will have a patch coming shortly. With the patch the ECN testcase between a neqo client and server succeeds.

Fixed with https://github.com/quinn-rs/quinn/pull/1765. Updated in https://github.com/mozilla/neqo/pull/1604 via https://github.com/mozilla/neqo/pull/1604/commits/a0d309385bfe6510593e74591f4f683ce0dd6ced.

**Comment by mxinden** [comment]
For convenience, I pushed my changes to https://github.com/mxinden/neqo/tree/feat-more-ecn-2. It includes:

- A rebase onto latest https://github.com/mozilla/neqo/pull/1604, including https://github.com/quinn-rs/quinn/pull/1765.
- A commit for each of the suggestions above.

File: neqo-transport/src/tracking.rs (hunk 3 above)
Comment (larseggert) [reply neqo-transport/src/tracking.rs:31]: I think @martinthomson meant something other by the "this" in his comment.

File: neqo-transport/src/tracking.rs (hunk 2 above)
Comment (larseggert) [reply neqo-transport/src/tracking.rs]: I had that originally, but was wondering if with multi-packet I/O we might want to increase by more than one. But I guess we can make that change at that point.

**Comment by larseggert** [comment]
See https://github.com/mozilla/neqo/pull/1678

**Pull request closed without merging**